modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/SkinMemoryLeakTest.java

```diff
@@ -82,6 +90,7 @@ public class SkinMemoryLeakTest {
      */
     @Test
     public void testMemoryLeakSameSkinClass() {
+        showControl(control, true);
         installDefaultSkin(control);
```

Comment:
no - here we test the isolated skin installation, no need for actually showing it. Please don't change the semantic of existing tests.

Reply:
So here we have a problem - I know that the memory leak is fixed (I've tested using VisualVM in a real application), but without this change, or without adding a StageLoader, the test fails.
Any suggestions?

Reply:
hmm, no direct idea but ... sounds like it's not fixed completely: first step would be to track down why it fails for this particular skin and what exactly is different from all other skins (that are fixed so far). If it turns out that there's something special, you can do the special memory testing in SkinCleanupTest (with documentation why that's required).

Reply:
btw: I would suggest that you believe the test :) Added complexity might cancel out an existing problem while the underlying issue still persists in some way ..

Reply:
Perhaps the test is too artificial, something is not being done correctly or exactly as in the real application?  Using StageLoader or showControl() hooks up the missing dependencies.

Reply:
> btw: I would suggest that you believe the test

Sorry, I trust VisualVM path to GC root.

Reply:
before:
<img width="939" alt="Screen Shot 2022-09-30 at 09 01 23" src="https://user-images.githubusercontent.com/107069028/193310763-90adba4e-ba88-45d3-9c24-d4b8b6738504.png">

Reply:
after:
<img width="940" alt="Screen Shot 2022-09-30 at 09 01 36" src="https://user-images.githubusercontent.com/107069028/193310795-48b2a72d-aad7-4c36-81d8-740f302fccb7.png">

Reply:
See https://github.com/andy-goryachev-oracle/Test/blob/main/src/goryachev/apps/LeakTest.java

Reply:
okay, not tested (so treat it as just a wild guess :) - there is this listener on the skinnable's sceneProperty: it's installed only when there's no scene and removed once the scene is set. Without showing, it looks like it's never removed (as it must!)

Unrelated: also look at your dispose: it's violating its contract (making SkinDisposeContractTest fail)

Reply:
Will definitely do!  Some tests were failing yesterday, until all is fixed - it's a draft PR :-)
Thank you so much, @kleopatra

Reply:
> Perhaps the test is too artificial, something is not being done correctly or exactly as in the real application? Using StageLoader or showControl() hooks up the missing dependencies.

one last time: there is _no_ such thing as a "too artificial" test - a class must _always_ fulfil its contract in whatever valid context. It's not enough to do so for some (or even the majority) of use-cases. Plus: logically, any assumption (like: there are no memory leaks) is invalidated by a single counter-example (like the valid test).

Have a nice weekend, I'm off now :)

Reply:
... as long as the context is *valid" :-)

thanks again, have a great weekend!

Reply:
You are right, @kleopatra  - this executeOnceWhenPropertyIsNonNull() is installing a listener.
perhaps we should add a similar functionality to LambdaMultiplePropertyChangeListenerHandler and actually install a WeakListener instead.

Reply:
Thanks again, @kleopatra 
With your permission, I'll add tests with and without scene property set.  Or do we want to keep the original set?

Reply:
> You are right, @kleopatra - this executeOnceWhenPropertyIsNonNull() is installing a listener. perhaps we should add a similar functionality to LambdaMultiplePropertyChangeListenerHandler and actually install a WeakListener instead.

I'd advise against doing this in an ad-hoc manner. Skins leak memory by adding (but not removing) all kinds of listeners or children to their controls. Some skins try to address this with weak listeners, or by manually removing listeners in the dispose() method. SkinBase attempts to make it easier for skins to not leak listeners with the `registerInvalidationListener` and `registerChangeListener` APIs, which remove an added listener when the skin is disposed (however, both of these methods only accept Consumers, not actual `InvalidationListener`/`ChangeListener` instances, making the API less useful than it could be).

But there are more places where memory is leaked:
1. adding, but not removing children to the control
2. adding EventHandlers/EventFilters to the control or to other objects (for example, MenuBarSkin adds EventHandlers to the control's scene)
3. hidden listeners like the one you discovered: Utils.executeOnceWhenPropertyIsNonNull
4. Label.setLabeLFor(control)
5. ListChangeListener, etc.

Most skins exhibit one or several of these memory leaks, which causes them to remain strongly referenced by the control even after being disposed. The problem here is that JavaFX doesn't offer good tools and APIs to address these challenges, so it's incredible easy to create a skin that leaks memory. In fact, it's so easy that most built-in skins do it (and these were created by the developers of JavaFX itself, so obviously it's a major problem if even those people get it wrong so often).

For me, this is a clear indication that we're dealing with an API problem first and foremost. I think there should be a better API that makes it easy for skins to interact with their associated controls, instead of hand-wiring hundreds of little skin-control dependencies (and sure enough, getting it wrong every so often).

Reply:
You are absolutely right!  There are so many places where memory leaks are being created.  My plan is to go through all the skins and take care of them all.

LambdaMultiplePropertyChangeListenerHandler is, I believe, a step in the right direction.  I wish there were a more uniform and public way to add various listeners, implemented as a public class - something that can be used to yank all listeners with a single method.

I've done something like that on the application level, see https://github.com/andy-goryachev/FxTextEditor/blob/master/src/goryachev/fx/FxDisconnector.java

What we could do (later) is to keep enhancing LambdaMultiplePropertyChangeListenerHandler by adding more ways of adding listeners, including full-blown change/invalidation listeners, and then rename it, move to a public util package and make it a part of the general purpose API.

Reply:
Sure, but I think this should be done with a clear vision of where we're going, and not just as a collection of ad-hoc helper methods. There are even more problems that a new API should at least consider:
Some skins change their behavior depending on the configuration of their associated control at the time when the skin is installed. For example, `MenuButtonSkinBase` adds a `MOUSE_PRESSED` handler only if the control didn't specify one for its onMousePressed property:

    if (control.getOnMousePressed() == null) {
        control.addEventHandler(MouseEvent.MOUSE_PRESSED, ...

But what happens if the `onMousePressed` property is set after the skin is installed?
Rather than a one-time check, cases like these should probably be a dynamic expression:

    when control.onMousePressedProperty().isNotNull() and skin.isInstalled()
        control.addEventHandler
    otherwise
        control.removeEventHandler

Reply:
> Thanks again, @kleopatra With your permission, I'll add tests with and without scene property set. Or do we want to keep the original set?

SkinMemoryLeakTest already has both methods, that is testing the replacement of a skin with/out residing in a scene .. no need for adding anything ;)

Reply:
> In fact, it's so easy that most built-in skins do it (and these were created by the developers of JavaFX itself, so obviously it's a major problem if even those people get it wrong so often).
> 

well, just guessing games here but: mine is that they didn't put too many thoughts into the possibility of replacing a skin and if they did, they actively prevented it - as seen in the implementation of TextAreaSkin.dispose when we started the current cleanup round:

    /** {@inheritDoc} */
    @Override public void dispose() {
        super.dispose();

        if (behavior != null) {
            behavior.dispose();
        }

        // TODO Unregister listeners on text editor, paragraph list
        throw new UnsupportedOperationException();
    }

Reply:
> You are right, @kleopatra - this executeOnceWhenPropertyIsNonNull() is installing a listener. perhaps we should add a similar functionality to LambdaMultiplePropertyChangeListenerHandler and actually install a WeakListener instead.

- yes, we need weakListeners (to wait for a not-null value to add the event handler) _and_ remove both the weakListener and the event handlers in dispose
- adding functionality to the LambdaHandler sounds like a good idea, but needs some thinking and should be done in a separate issue which also adds delegate methods to expose the new functionality in SkinBase for use in sub classes

For this issue, you might simply inline the utility method and register the listeners using skinbase api (didn't try, it's your issue :), see TableRowSkin for a similar pattern 

This conditional adding of handlers/listeners is needed if we have to support path properties (also f.i. when listening to selectedXX in any of the skins with selectionModels as properties). Which basically should be supported by base - but probably without any chance to ever get them ;)

Reply:
ListenerHelper - replacement for LambdaMultiplePropertyChangeListenerHandler .
Please take a look at https://github.com/openjdk/jfx/pull/908

Reply:
> Sure, but I think this should be done with a clear vision of where we're going, and not just as a collection of ad-hoc helper methods. There are even more problems that a new API should at least consider: Some skins change their behavior depending on the configuration of their associated control at the time when the skin is installed. For example, `MenuButtonSkinBase` adds a `MOUSE_PRESSED` handler only if the control didn't specify one for its onMousePressed property:
> 
> ```
> if (control.getOnMousePressed() == null) {
>     control.addEventHandler(MouseEvent.MOUSE_PRESSED, ...
> ```
> 
> But what happens if the `onMousePressed` property is set after the skin is installed? Rather than a one-time check, cases like these should probably be a dynamic expression:
> 
> ```
> when control.onMousePressedProperty().isNotNull() and skin.isInstalled()
>     control.addEventHandler
> otherwise
>     control.removeEventHandler
> ```

I think you meant `isNull()` instead of `isNotNull()`:

    when control.onMousePressedProperty().isNull() and skin.isInstalled()
        control.addEventHandler
    otherwise
        control.removeEventHandler

So... I was curious, and was checking to see if this can be done with fluent bindings these days, and I can get quite close. First assume we have a convenient property in all skins (or everything that is `SkinBase`) that reflects whether or not the skin is currently installed or active:

        BooleanProperty installed = new SimpleBooleanProperty();

You can then create a condition:

        ObservableValue<Boolean> noMouseAndInstalled = installed
            .flatMap(b -> b ? control.onMousePressedProperty() : installed)
            .map(x -> false)  // ---+ use case for isNotEmpty
            .orElse(true);    // --/

Or (if `isNotEmpty` already exists): 

        ObservableValue<Boolean> noMouseAndInstalled = installed
            .flatMap(b -> b ? control.onMousePressedProperty() : installed)
            .isNotEmpty();

It's then a simple matter of adding a listener to this result to either add or remove the event handler:

         installed
            .flatMap(b -> b ? control.onMousePressedProperty() : installed)
            .isNotEmpty();
            .addListener((obs, old, active) -> {
               if(active) control.addEventHandler(...);
               else control.removeEventHandler(...);
            });

One of the nice things here is that when `installed` is `false`, the listener on `control.onMousePressedProperty()` is automatically removed.  This is because the `flatMap` will switch to listening to `installed`.  This should mean that it will automatically be eligible for GC when `installed` becomes `false`.

The second nice thing is that you can do this in the constructor already, no need to do this in the `install` callback.  If you do it in the callback, then make sure to do it before `installed` is set to `true` as otherwise the change listener won't fire for the first change (another reason to have value listeners...)

If this becomes a common idiom, we could introduce a short circuiting `and` construct so the `flatMap` can be expressed easier (the short circuiting part will ensure the onMousePressedProperty won't be observed if installed is `false`):

      installed
          .and(control.onMousePressedProperty().isNotEmpty())
          .addListener( ... );